### DISCUSSION TITLE
Single-file Haskell scripts with Cabal directives

### DISCUSSION THREAD

**Discussion by chris-martin** [haskell/haskell-language-server#3572] [General]
[From the Cabal documentation](https://cabal.readthedocs.io/en/3.10/getting-started.html#run-a-single-file-haskell-script):

> Cabal also enables us to run single-file Haskell scripts without creating a project directory or .cabal file. The cabal directives are placed in the file within a comment.
>
> ```haskell
> #!/usr/bin/env cabal
> {- cabal:
> build-depends: base, split
> -}
> 
> import Data.List.Split (chunksOf)
> 
> main :: IO ()
> main = getLine >>= print . chunksOf 3
> ```

If I follow this example, I can indeed run this script at the command line. However, Haskell language server gives an error:

```
[{
	"resource": "[...].hs",
	"owner": "Haskell (invalid)",
	"severity": 8,
	"message": "Could not find module ‘Data.List.Split’\nIt is not a module in the current program, or in any known package.",
	"source": "not found",
	"startLineNumber": 6,
	"startColumn": 8,
	"endLineNumber": 6,
	"endColumn": 23
}]
```

Do I correctly infer that HLS does not support this feature? Are there any plans to?

**Comment by fendor** [comment]
> Do I correctly infer that HLS does not support this feature?

Yes, that's correct.

> Are there any plans to?

Of course, that'd be great. In theory, it is pretty simple to support, we *just* need a way to find out how to compile this file. 

In an experiment, I found out that the following setup seems to be working, would you mind confirming?

Add a hie.yaml like this:
```yaml
cradle:
  cabal:
```

and an empty `cabal.project` file.

Changes to the script part are not picked up, so you have to restart HLS in that case.

**Reply by chris-martin** [reply]
Can confirm, that does work! (GHC 9.2.8, cabal-install 3.10.1.0, haskell-language-server 2.0.0.0)

**Comment by Anrock** [comment]
Is this still not supported? Is there a tracking issue?

**Reply by fendor** [reply]
https://github.com/haskell/haskell-language-server/issues/111
